sms_analysis.ipynb

```diff
@@ -779,6 +779,149 @@
     "    top_n=widgets.IntSlider(min=15, max=100, step=1, value=5, description='Max words to show:')\n",
     ")"
    ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {
+    "collapsed": false
+   },
+   "outputs": [],
+   "source": [
+    "from sklearn.cluster import KMeans, MiniBatchKMeans\n",
+    "from sklearn.decomposition import TruncatedSVD\n",
+    "from sklearn.manifold import TSNE\n",
+    "\n",
+    "def _top_words_by_cluster_from_tfidf(\n",
+    "    cluster_id,\n",
+    "    x_data_tfidf_sender,\n",
+    "    tfidf_per_sender_cluster,\n",
+    "    top_n=15,\n",
+    "):\n",
+    "    \"\"\"\n",
+    "    Returns a dataframe of the top words for each cluster by their TFIDF score.\n",
+    "    \n",
+    "    To determine the \"top\", we look at one cluster's TFIDF - avg(other clusters' TFIDFs)\n",
+    "    \n",
+    "    Args:\n",
+    "        cluster_id: The cluster we want to find the top words for\n",
+    "        x_data_tfidf_sender: TFIDF matrix with as many rows as entries in tfidf_per_sender_cluster\n",
+    "        tfidf_per_sender_cluster: Cluster assignment for each entry in x_data_tfidf_sender\n",
+    "        top_n: Number of top words per cluster to include in the result\n",
+    "    \"\"\"\n",
+    "    this_cluster_indicies = np.argwhere(tfidf_per_sender_cluster == cluster_id)\n",
+    "    other_cluster_indicies = np.argwhere(tfidf_per_sender_cluster != cluster_id)\n",
+    "    this_cluster_records = x_data_tfidf_sender[this_cluster_indicies.flatten()]\n",
+    "    other_cluster_records = x_data_tfidf_sender[other_cluster_indicies.flatten()]\n",
+    "    \n",
+    "    mean_this_cluster = np.asarray(this_cluster_records.mean(axis=0)).reshape(-1)\n",
+    "    mean_other_cluster = np.asarray(other_cluster_records.mean(axis=0)).reshape(-1)\n",
+    "    \n",
+    "    difference = mean_this_cluster - mean_other_cluster\n",
+    "    most_different_indicies = difference.argsort()\n",
+    "    return most_different_indicies[-top_n:]\n",
+    "\n",
+    "def _tfidf_by_sender(slice_of_texts_df, min_texts_required=100):\n",
```

Comment:
What is this method really doing?  Why are you talking about "years" in the comments?  Can't you leverage the existing _get_tfidf_record_for_contact() I wrote above?  This will also prevent you from re-vectorizing the data as the TFIDF matrix already exists.